module/zfs/dsl_dataset.c

```diff
@@ -1757,6 +1757,20 @@ dsl_dataset_snapshot_sync_impl(dsl_dataset_t *ds, const char *snapname,
 		}
 	}
 
+	/*
+	 * We are not allowed to dirty a filesystem when done receiving
+	 * a snapshot. In this case the flag SPA_FEATURE_LARGE_BLOCKS will
+	 * not be set and a subsequent encrypted raw send will fail. Hence
+	 * activate this feature if needed here.
+	 */
+	spa_feature_t f = SPA_FEATURE_LARGE_BLOCKS;
+	if (zfeature_active(f, ds->ds_feature_activation[f]) &&
+	    !(zfeature_active(f, ds->ds_feature[f]))) {
+		dsl_dataset_activate_feature(dsobj, f,
+		    ds->ds_feature_activation[f], tx);
+		ds->ds_feature[f] = ds->ds_feature_activation[f];
+	}
```

Comment:
It seems to me we should be processing `ds->ds_feature_activation[f]` for all of the features, like at the end of `dsl_dataset_sync()`.  Otherwise, we still may not correctly enable the `SPA_FEATURE_BLAKE3`, `SPA_FEATURE_SHA512`, `SPA_FEATURE_SKEIN`, `SPA_FEATURE_EDONR`, and `SPA_FEATURE_ZSTD_COMPRESS` features.  Although, I haven't tested this so I may be missing something.

Reply:
Makes sense, I modified this as you suggested.